File: fpinjava-parent/fpinjava-lists-exercises/src/main/java/com/fpinjava/lists/exercise05_22/Filter.java
package com.fpinjava.lists.exercise05_22;

import com.fpinjava.common.Function;
import com.fpinjava.lists.exercise05_21.List;

public class Filter {

  public static <A> List<A> filterViaFlatMap(List<A> list, Function<A, Boolean> p) {
    return list.flatMap(x -> p.apply(x) ? List.list(x) : List.list());
  }
}

File: fpinjava-parent/fpinjava-lists-exercises/src/main/kotlin/com/fpinkotlin/lists/exercise05_22/Filter.kt
package com.fpinkotlin.lists.exercise05_22

import com.fpinkotlin.lists.exercise05_21.List
import com.fpinkotlin.lists.exercise05_21.list

fun <A> filterViaFlatMap(list: List<A>, p: (A) -> Boolean): List<A> =
		list.flatMap { x -> if (p(x)) list(x) else list<A>() }
